### PR TITLE
test: add CountingReporter integration tests across different scenarios

### PR DESCRIPTION
**Pull request by chiinlquah**
## What changes are proposed in this pull request?

Adds integration tests for `CountingReporter` metrics across all major snapshot-loading and scan code paths. All tests use only public kernel APIs.

**Test organization** (`kernel/tests/metrics/`):
- `snapshot_load.rs` -- 8 scenarios covering delta-only, V1 checkpoint (with/without tail), log compaction, CRC fast-path, stale CRC, stale checkpoint, and on-demand `get_domain_metadata`
- `scan.rs` -- `scan.execute()` parquet data-file reads (isolated from snapshot-build I/O)
- `main.rs` -- shared helpers and crate root

**CountingReporter cleanup** (`test-utils/src/counting_reporter.rs`):
- Introduces `RelaxedCounter` wrapper over `AtomicU64` with `inc()`, `add()`, `get()`, `reset()` methods, replacing verbose `fetch_add(..., Ordering::Relaxed)` / `load(..., Ordering::Relaxed)` / `store(0, ...)` calls throughout
- All `CountingReporter` fields switch from raw `AtomicU64` to `RelaxedCounter`
- `RelaxedCounter` is re-exported from `test_utils`

## How was this change tested?

`cargo nextest run -p delta_kernel --test metrics --all-features` -- all 9 tests pass.

### CODE REVIEW

File: kernel/tests/metrics.rs [hunk 1]
```diff
@@ -0,0 +1,344 @@
...
+/// A snapshot built from raw JSON commits -- no checkpoint, no CRC, no compaction --
+/// reports exactly the commit file count and triggers one JSON read call covering all
+/// commit files.
+#[tokio::test]
+async fn delta_only_snapshot_emits_expected_metrics() -> Result<(), Box<dyn std::error::Error>> {
+    let store = Arc::new(InMemory::new());
+    let table_root = "memory:///";
+
+    // commit 0: protocol + metadata; commit 1: add one file
+    add_commit(
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: nit: please use kernel public apis to create tables :D

File: kernel/tests/metrics.rs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+
+    // commit 1: insert data
+    let snap0 = Snapshot::builder_for(table_url.clone()).build(setup_engine.as_ref())?;
+    let result = insert_data(
+        snap0,
+        &setup_engine,
+        vec![Arc::new(Int32Array::from(vec![1]))],
+    )
+    .await?;
+    let CommitResult::CommittedTransaction(committed) = result else {
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: you can use `result.unwrap_committed`

File: kernel/tests/metrics.rs [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+
+    // commit 0: protocol + metadata; commit 1: add one file
+    add_commit(
+        table_root,
+        store.as_ref(),
+        0,
+        actions_to_string(vec![TestAction::Metadata]),
+    )
+    .await?;
+    add_commit(
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: i also encourage us to use kernel public APIs for inserts -- if this is not simple + easy, we need to develop the right test utilities to _always_ favor kernel production code (production APIs) over these mock test helpers!

File: kernel/src/engine/default/mod.rs [hunk 4]
```diff
@@ -87,6 +87,63 @@ impl<T: Send + 'static, E: executor::TaskExecutor> Iterator for BlockingStreamIt
...
+    }
+}
+
+impl Iterator for ReadMetricsIterator {
+    type Item = crate::DeltaResult<Box<dyn crate::EngineData>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let item = self.inner.next();
+        if item.is_none() {
+            self.emit_once();
```
Comment (scovich) [inline kernel/src/engine/default/mod.rs:135]: out of curiosity, are there scenarios where the iterator would not drop ~immediately after being exhausted, so that it matters to emit eagerly here instead?

File: test-utils/src/counting_reporter.rs [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+            MetricEvent::StorageListCompleted { num_files, .. } => {
+                self.list_calls.fetch_add(1, Ordering::Relaxed);
+                self.list_files_seen.fetch_add(num_files, Ordering::Relaxed);
+            }
+            MetricEvent::StorageReadCompleted {
+                num_files,
+                bytes_read,
+                ..
+            } => {
+                self.storage_read_calls.fetch_add(1, Ordering::Relaxed);
```
Comment (scovich) [inline test-utils/src/counting_reporter.rs]: This fetch_add with relaxed ordering shows up a _lot_. Should we define a newtype wrapper for it?

File: kernel/src/engine/default/mod.rs (hunk 4 above)
Comment (chiinlquah) [reply kernel/src/engine/default/mod.rs:135]: Thinking out loud - ReadMetricsIterator can be wrapped inside an outer iterator or held by other components and when the inner iterator is exhausted (next() returns None), the outer wrapper or component can still be alive in the caller's hand for an arbitrary amount of time before its own Drop fires. 

The eager emit_once() in next() means the metric fires as soon as the last item is produced, rather than waiting for the wrapper to go out of scope.

File: kernel/tests/metrics/snapshot_load.rs
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs:1]: let's do something like kernel/tests/metrics/snapshot_load.rs ?

because: snapshot.get_X and transaction and scan etc in the future will be separate?

File: kernel/tests/metrics/snapshot_load.rs [hunk 6]
```diff
@@ -0,0 +1,584 @@
...
+    assert_eq!(reporter.commit_files.get(), 2);
+    assert_eq!(reporter.checkpoint_files.get(), 0);
+    assert_eq!(reporter.compaction_files.get(), 0);
+
+    // Both commit files read in a single JSON call
+    assert_eq!(reporter.json_read_calls.get(), 1);
+    assert_eq!(reporter.json_files_read.get(), 2);
+    assert_eq!(reporter.parquet_read_calls.get(), 0);
+
+    // Exactly one listing covering exactly the 2 commit files
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs:54]: @lorenarosati -- will you able to add kernel/tests/metrics/time_travel.rs with your backwards log listing work?

File: kernel/tests/metrics/snapshot_load.rs [hunk 7]
```diff
@@ -0,0 +1,584 @@
...
+    assert_eq!(reporter.list_calls.get(), 1);
+    assert_eq!(reporter.list_files_seen.get(), 2);
+
+    Ok(())
+}
+
+// ---------------------------------------------------------------------------
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs]: nit: @chiinlquah -- can you chat with the team about whether we should use
`// ----` vs `// ====`

this is small and silly, but we _should_ be on the same page about it. we shouldn't be incosistent

File: kernel/tests/metrics/snapshot_load.rs [hunk 8]
```diff
@@ -0,0 +1,584 @@
...
+            snap,
+            &setup_engine,
+            vec![Arc::new(Int32Array::from(vec![val]))],
+        )
+        .await?;
+    }
+
+    // Write a compacted log file covering versions 0-2 using the public API
+    let snap2 = Snapshot::builder_for(table_url.clone()).build(setup_engine.as_ref())?;
+    let mut writer = snap2.log_compaction_writer(0, 2)?;
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs:162]: TIL we have a log compaction writer

File: kernel/tests/metrics/snapshot_load.rs
Comment (scottsand-db) [reply kernel/tests/metrics/snapshot_load.rs:1]: actually, i'm worried that when we add tests for `get_X` (domain metadata, timestamp (ICT), set transaction, table file stats, etc) that we will be repeating all of our table setup here, eh?

@DrakeLin with your table builder -- will we be able to easily inline a table setup definition? like
- table_only_deltas
- table_checkpoint
- table_checkpoint_with_deltas
- etc.

File: kernel/tests/metrics.rs [hunk 9]
```diff
@@ -0,0 +1,584 @@
...
+    assert_eq!(reporter.json_read_calls.get(), 1);
+    assert_eq!(reporter.json_files_read.get(), 2);
+    // CRC is loaded from storage as a fallback (P+M not found in the tail commits)
+    assert!(reporter.storage_read_calls.get() >= 1);
+
+    Ok(())
+}
+
+// ---------------------------------------------------------------------------
+// Scenario 7: stale checkpoint (checkpoint with many tail commits)
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: nit: wouldn't really call it `stale` -- this is common, expected, valid

File: kernel/tests/metrics/snapshot_load.rs [hunk 10]
```diff
@@ -0,0 +1,584 @@
...
+    assert_eq!(reporter.json_files_read.get(), 3);
+    assert_eq!(reporter.parquet_read_calls.get(), 1);
+    assert!(reporter.json_bytes_read.get() > 0);
+    assert!(reporter.parquet_bytes_read.get() > 0);
+
+    Ok(())
+}
+
+// ---------------------------------------------------------------------------
+// Scenario 8: on-demand domain metadata query incurs additional log replay
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs:352]: ah, see my comments above -- this will be different for different tables!

File: kernel/tests/metrics.rs [hunk 11]
```diff
@@ -0,0 +1,584 @@
...
+        1,
+        "get_domain_metadata replays the log, incurring one additional JSON read call"
+    );
+    assert!(reporter.json_files_read.get() > 0);
+
+    Ok(())
+}
+
+// ---------------------------------------------------------------------------
+// Scenario 9: snapshot.transaction() on a clustered table loads clustering metadata
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: definitely different file

File: kernel/tests/metrics.rs [hunk 12]
```diff
@@ -0,0 +1,584 @@
...
+    assert_eq!(reporter.parquet_files_read.get(), 2);
+    assert!(reporter.parquet_bytes_read.get() > 0);
+    // scan.execute() does its own log replay for Add/Remove scan metadata
+    assert_eq!(reporter.json_read_calls.get(), 1);
+
+    Ok(())
+}
+
+// ---------------------------------------------------------------------------
+// Scenario 11: incremental snapshot update replays only new tail commits
```
Comment (scottsand-db) [inline kernel/tests/metrics.rs]: include more context about their log segments? lots of wonky things go on here ...

e.g. if, when looking from the prev snapshot's checkpoint, we find a new checkpoint, for some (silly) reason we just totally ignore the prev snapshot and builda brand new one! (nothing is incremental...)

I don't think we should check this (test) in. We should instead analyze the code and identify all the interesting test cases. tests/metrics/incremental_snapshot.rs should perhaps be its own file? I defer to you

File: kernel/tests/metrics/snapshot_load.rs (hunk 7 above)
Comment (chiinlquah) [reply kernel/tests/metrics/snapshot_load.rs]: Just scanned through the codebase and seems like the convention is ===, let me update it here and update claude.md file too.

File: kernel/tests/metrics/snapshot_load.rs (hunk 10 above)
Comment (chiinlquah) [reply kernel/tests/metrics/snapshot_load.rs:352]: Added the different dimensions, 
```
  - Delta-only table → full JSON log replay
  - Table with CRC at target version → read from CRC, no replay
  - Table with checkpoint → parquet read + tail replay
  - Table with CRC at prior version → partial replay + CRC fallback
  - ```

File: kernel/tests/metrics.rs (hunk 12 above)
Comment (chiinlquah) [reply kernel/tests/metrics.rs]: moved to incremental_snapshot.rs and added the below cases
```
  1. No new commits → returns existing snapshot, no LogSegmentLoaded
  2. New commits only → incremental tail replay (1 commit)
  3. Multiple new commits → all read in single JSON call
  4. New checkpoint found → full rebuild (documents the metrics gap where LogSegmentLoaded is not emitted)
  5. Compaction overlapping old version → conservatively filtered out

```

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 13]
```diff
@@ -0,0 +1,248 @@
...
+//!
+//! `Snapshot::builder_from(existing)` has several behavioral branches depending on what
+//! the log listing finds between the existing snapshot's state and the target version:
+//!
+//! - **No new commits (no version requested):** returns existing snapshot unchanged, no
+//!   new metrics emitted (only SnapshotCompleted from the builder).
+//! - **New commits only (no new checkpoint):** incrementally replays only the net-new
+//!   commits for Protocol/Metadata changes, then merges with the existing log segment.
+//!   `LogSegmentLoaded` reports the net-new commit count.
+//! - **New checkpoint found:** ignores the existing snapshot entirely and constructs a
```
Comment (rliao147) [inline kernel/tests/metrics/incremental_snapshot.rs]: I'm adding some code to change the behavior in https://github.com/delta-io/delta-kernel-rs/issues/2218. Would be cool to see the performance improvements with this metrics PR.

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 14]
```diff
@@ -0,0 +1,248 @@
...
+
+// ============================================================================
+// No new commits -- returns existing snapshot unchanged
+// ============================================================================
+
+/// When no new commits exist after the existing snapshot, `builder_from` returns the
+/// same snapshot. `SnapshotCompleted` is still emitted (from the builder) but no
+/// `LogSegmentLoaded` is emitted because no new log segment was constructed.
+#[tokio::test]
+async fn no_new_commits_returns_existing_snapshot_no_log_segment_metric() -> DeltaResult<()> {
```
Comment (rliao147) [inline kernel/tests/metrics/incremental_snapshot.rs]: nit: I know that there's already a `#[tokio::test]` macro, but should we also prefix these tests with `test_`? I see a majority of tests in the codebase also have the same `test_` prefix. If so, we can also add it to `/kernel-review.`

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 15]
```diff
@@ -0,0 +1,248 @@
...
+
+// ============================================================================
+// New checkpoint found -- full rebuild (not incremental)
+// ============================================================================
+
+/// When a new checkpoint is discovered after the existing snapshot, `builder_from`
+/// ignores the existing snapshot and builds a fresh one from the new checkpoint.
+/// Nothing is incremental -- the existing snapshot is completely discarded.
+///
+/// This is the "wonky" case Scott flagged: even if the existing snapshot had most of
```
Comment (rliao147) [inline kernel/tests/metrics/incremental_snapshot.rs]: Let's remove this internal dialogue. But agreed that the case is wonky :)

Comment (rliao147) [reply kernel/tests/metrics/incremental_snapshot.rs]: I'll most likely change this case to only rebuild if it's absolutely necessary.

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 16]
```diff
@@ -0,0 +1,248 @@
...
+    // Incremental update from v1 -- discovers checkpoint at v3
+    let (measure_engine, reporter) = measuring_engine(Arc::new(LocalFileSystem::new()));
+    let updated = Snapshot::builder_from(existing_snap).build(&measure_engine)?;
+
+    assert_eq!(updated.version(), 4);
+    assert_eq!(reporter.snapshot_completions.get(), 1);
+
+    // The incremental-with-new-checkpoint path does NOT emit LogSegmentLoaded (the log
+    // segment is built via try_new, not for_snapshot). This is a metrics gap -- but we
+    // document the current behavior here so any future change to emit it is intentional.
```
Comment (rliao147) [inline kernel/tests/metrics/incremental_snapshot.rs]: should we file an issue?

File: kernel/tests/metrics/scan.rs [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+    // Reset after snapshot build to isolate scan I/O
+    reporter.reset();
+
+    let engine: Arc<dyn Engine> = Arc::new(engine);
+    let mut batches_seen = 0usize;
+    for result in snap.scan_builder().build()?.execute(engine)? {
+        result?;
+        batches_seen += 1;
+    }
+    assert!(batches_seen > 0, "scan should return rows");
```
Comment (rliao147) [inline kernel/tests/metrics/scan.rs]: nit: `assert_eq!(batches_seen, 2, "scan should return one batch per data file");`

File: kernel/tests/metrics/incremental_snapshot.rs (hunk 14 above)
Comment (chiinlquah) [reply kernel/tests/metrics/incremental_snapshot.rs]: seems like the convention for integration tests is not to have test_ prefix but just have #test or #tokio::test tag.

File: kernel/src/snapshot.rs [hunk 18]
```diff
@@ -288,6 +280,20 @@ impl Snapshot {
             .ascending_compaction_files
             .retain(|log_path| old_version < log_path.version);
 
+        // Emit LogSegmentLoaded now that we know the net-new commit and compaction counts
+        // (after the retain above removes overlap with the existing snapshot).
```
Comment (kyli87) [inline kernel/src/snapshot.rs]: This feels inconsistent/confusing/wrong to me for a few reasons:
1. There are still other early return cases above (e.g line 249) where we would return a loaded LogSegment but failed to emit the event.
2. The exact semantics of LogSegmentLoaded are not clearly defined. Should the metrics reflect how many files were listed or how many were ultimately used in a resulting snapshot? I think you've chosen the latter here, but then what should happen on line 327 when we merge LogSegments into yet another new LogSegment? Should we emit another event?
3. Here we emit the LogSegmentLoaded event within a `Snapshot` constructor but in the other case we emit it in `LogSegment::for_snapshot`. This is probably why we end up with the inconsistent/confusing behavior I described above.

My recommendation is to first define exactly what LogSegmentLoaded means (does it apply to snapshot creation or anytime we list files or something else)? And then pick a consistent layer to emit the event.

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 19]
```diff
@@ -0,0 +1,297 @@
+//! Incremental snapshot update metrics tests.
+//!
+//! `Snapshot::builder_from(existing)` applies a simple heuristic to minimize I/O when
```
Comment (kyli87) [inline kernel/tests/metrics/incremental_snapshot.rs]: nit: This comment seems out of place or I'm not clear on the relevancy to this test file? I think just a line like "the goal of this test file to to verify the I/O expectations in these scenarios" would go a long way.

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 20]
```diff
@@ -0,0 +1,297 @@
...
+//! - **No new commits:** returns the existing snapshot unchanged; `SnapshotCompleted` is
+//!   emitted but no `LogSegmentLoaded` (no log replay occurred).
+//! - **New commits only (no new checkpoint):** incrementally replays only the net-new
+//!   commits for Protocol/Metadata changes, then merges with the existing log segment.
+//!   `LogSegmentLoaded` is emitted with the net-new commit count.
+//! - **New checkpoint found:** ignores the existing snapshot entirely and constructs a
+//!   fresh snapshot from the new checkpoint (plus any tail commits after it). This path
+//!   is equivalent to a full `builder_for` build -- nothing is incremental.
+//!   `LogSegmentLoaded` is not emitted (the checkpoint path uses `LogSegment::try_new`
+//!   rather than `LogSegment::for_snapshot`).
```
Comment (kyli87) [inline kernel/tests/metrics/incremental_snapshot.rs]: nit: prefer avoiding references to existing code/functions since it's describing the implementation too much (which could easily change)

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 21]
```diff
@@ -0,0 +1,297 @@
...
+        delta_kernel::engine::default::DefaultEngine<
+            delta_kernel::engine::default::executor::tokio::TokioBackgroundExecutor,
+        >,
+    >,
+    start_val: i32,
+    count: i32,
+) -> DeltaResult<()> {
+    for val in start_val..(start_val + count) {
+        let snap = Snapshot::builder_for(table_url.clone()).build(engine.as_ref())?;
+        let _ = insert_data(snap, engine, vec![Arc::new(Int32Array::from(vec![val]))]).await?;
```
Comment (kyli87) [inline kernel/tests/metrics/incremental_snapshot.rs]: nit: I think it'd be better to use `.unwrap_committed()` and then get the next snapshot from `CommittedTransaction::post_commit_snapshot`. That way you fail fast on any commit errors and also avoid having to rebuild the snapshot each time.

File: kernel/tests/metrics/incremental_snapshot.rs [hunk 22]
```diff
@@ -0,0 +1,297 @@
...
+use delta_kernel::object_store::path::Path;
+use delta_kernel::object_store::ObjectStore as _;
+use delta_kernel::transaction::create_table::create_table;
+use delta_kernel::{DeltaResult, Snapshot};
+use test_utils::{insert_data, test_table_setup_mt};
+use url::Url;
+
+/// Create an in-memory table with `num_inserts` commits after the initial create-table
+/// commit. Returns `(table_url, setup_engine, store)`. Table is at version `num_inserts`.
+async fn setup_in_memory_table(
```
Comment (kyli87) [inline kernel/tests/metrics/incremental_snapshot.rs]: These helpers could go in `main.rs` instead?

File: kernel/tests/metrics/main.rs [hunk 23]
```diff
@@ -0,0 +1,88 @@
...
+//! All tests use only public kernel APIs (`create_table`, `Transaction`, `write_parquet`,
+//! `log_compaction_writer`, `scan.execute()`) to ensure that the metrics surface is
+//! verified through the same code paths a real engine connector would exercise.
+//!
+//! Tests are organized by concern:
+//! - [`snapshot_load`]: snapshot-loading scenarios (delta-only, checkpoint, compaction, CRC,
+//!   and on-demand API calls like `get_domain_metadata`)
+//! - [`scan`]: scan execution scenarios (`scan.execute()` parquet data-file reads)
+//! - [`incremental_snapshot`]: incremental update scenarios via `Snapshot::builder_from`
+
```
Comment (kyli87) [inline kernel/tests/metrics/main.rs:16]: Why is this file called main.rs and not mod.rs?

File: kernel/tests/metrics/scan.rs [hunk 24]
```diff
@@ -0,0 +1,80 @@
...
+use delta_kernel::committer::FileSystemCommitter;
+use delta_kernel::engine::default::DefaultEngineBuilder;
+use delta_kernel::object_store::memory::InMemory;
+use delta_kernel::transaction::create_table::create_table;
+use delta_kernel::{DeltaResult, Engine, Snapshot};
+use test_utils::insert_data;
+use url::Url;
+
+// ============================================================================
+// Scenario 10: scan.execute() contributes parquet data-file reads
```
Comment (kyli87) [inline kernel/tests/metrics/scan.rs]: nit: Numbering of scenarios should restart in each file...?

File: kernel/tests/metrics/scan.rs [hunk 25]
```diff
@@ -0,0 +1,80 @@
...
+    // Two inserts, each writing one parquet data file
+    for val in [1i32, 2] {
+        let snap = Snapshot::builder_for(table_url.clone()).build(setup_engine.as_ref())?;
+        let _ = insert_data(
+            snap,
+            &setup_engine,
+            vec![Arc::new(Int32Array::from(vec![val]))],
+        )
+        .await?;
+    }
```
Comment (kyli87) [inline kernel/tests/metrics/scan.rs]: could this just use the `setup_table` helper you defined in `incremental_snapshot` tests?

File: kernel/tests/metrics/snapshot_load.rs [hunk 26]
```diff
@@ -0,0 +1,407 @@
...
+    let _ = create_table("memory:///", simple_schema(), "Test/1.0")
+        .build(setup_engine.as_ref(), Box::new(FileSystemCommitter::new()))?
+        .commit(setup_engine.as_ref())?;
+    let snap0 = Snapshot::builder_for(table_url.clone()).build(setup_engine.as_ref())?;
+    let _ = insert_data(
+        snap0,
+        &setup_engine,
+        vec![Arc::new(Int32Array::from(vec![1]))],
+    )
+    .await?;
```
Comment (kyli87) [inline kernel/tests/metrics/snapshot_load.rs]: same comment - resue setup_table helper?

File: kernel/tests/metrics/snapshot_load.rs [hunk 27]
```diff
@@ -0,0 +1,407 @@
...
+    assert_eq!(reporter.json_read_calls.get(), 1);
+    assert_eq!(reporter.json_files_read.get(), 1);
+    assert_eq!(reporter.parquet_read_calls.get(), 1);
+    assert_eq!(reporter.parquet_files_read.get(), 1);
+    // On Windows (NTFS), listing a file written moments earlier can return size=0 because
+    // the OS has not yet flushed size metadata to the directory entry. bytes_read is sourced
+    // from these FileMeta::size values, so it can be 0 even when real I/O occurred.
+    #[cfg(not(windows))]
+    assert!(reporter.json_bytes_read.get() > 0);
+    #[cfg(not(windows))]
```
Comment (kyli87) [inline kernel/tests/metrics/snapshot_load.rs:102]: Just curious how you determined this? Do we somehow build/verify on Windows in CI?

File: kernel/tests/metrics/snapshot_load.rs [hunk 28]
```diff
@@ -0,0 +1,407 @@
...
+
+    for val in [1i32, 2] {
+        let snap = Snapshot::builder_for(table_url.clone()).build(setup_engine.as_ref())?;
+        let _ = insert_data(
+            snap,
+            &setup_engine,
+            vec![Arc::new(Int32Array::from(vec![val]))],
+        )
+        .await?;
+    }
```
Comment (kyli87) [inline kernel/tests/metrics/snapshot_load.rs]: Use setup_table?

File: kernel/tests/metrics/main.rs (hunk 23 above)
Comment (chiinlquah) [reply kernel/tests/metrics/main.rs:16]: I believe that's the convention in rust. Integration test directories uses main.rs as crate root.

File: kernel/tests/metrics/incremental_snapshot.rs (hunk 21 above)
Comment (chiinlquah) [reply kernel/tests/metrics/incremental_snapshot.rs]: Good suggestion. Changed.

File: kernel/tests/metrics/main.rs (hunk 23 above)
Comment (kyli87) [reply kernel/tests/metrics/main.rs:16]: I think technically main.rs would be appropriate if there are tests in this file. But if its just for helper functions, we should use mod.rs. Because otherwise cargo will assume this is a test crate and try to run it. See: https://doc.rust-lang.org/book/ch11-03-test-organization.html#submodules-in-integration-tests

File: kernel/src/snapshot.rs (hunk 18 above)
Comment (chiinlquah) [reply kernel/src/snapshot.rs]: I think my intention is to list only the deltas from the previous snapshot. The 243 is an edge case where we just re consume everything starting from the new checkpoint.

I would propose that we keep this as is and add a new report for LogSegmentLoaded for the above case. With that case, we can also easily differentiate the different reports (where if num_checkpoint_files = 0, then we know that its a pure incr snapshot construction vs full snapshot construction).

The definition for LogSegmentLoaded would just be the different files applied to the Snapshot (new and existing).

File: kernel/tests/metrics/snapshot_load.rs (hunk 27 above)
Comment (chiinlquah) [reply kernel/tests/metrics/snapshot_load.rs:102]: Yea we run the tests through different OS in the CI jobs (macOS/windows/ubuntu).

File: kernel/tests/metrics/main.rs (hunk 23 above)
Comment (chiinlquah) [reply kernel/tests/metrics/main.rs:16]: But regardless we will still need main.rs for the test create entry point (defines all the submodules scna, snapshot load etc). Do you propose to keep main.rs just for that purpose and move the test helpers into mod.rs?

Comment (kyli87) [reply kernel/tests/metrics/main.rs:16]: I suspect you don't need main.rs? from my understanding any file under `tests` will be treated as a test main file. Can you try just renaming to mod.rs and verifying if tests are still being run?

File: kernel/src/snapshot.rs (hunk 18 above)
Comment (kyli87) [reply kernel/src/snapshot.rs]: I think it's fine to decouple this discussion from this PR, but we should file an issue and link a TODO in the code to revisit. I think just patching this function with additional LogSegmentLoaded events will be brittle.

File: kernel/tests/metrics/main.rs (hunk 23 above)
Comment (chiinlquah) [reply kernel/tests/metrics/main.rs:16]:
```
error: no test target named `metrics` in `delta_kernel` package
```
it wasn't able to find the metrics test crate

File: kernel/src/snapshot.rs [hunk 29]
```diff
@@ -288,6 +280,20 @@ impl Snapshot {
...
+        let reporter = engine.get_metrics_reporter();
+        reporter.inspect(|r| {
```
Comment (kyli87) [inline kernel/src/snapshot.rs]: This change should also be moved to your child PR with the incremental snapshot tests?

Comment (chiinlquah) [reply kernel/src/snapshot.rs]: yup

File: kernel/tests/metrics/main.rs [hunk 30]
```diff
@@ -0,0 +1,145 @@
...
+/// Create a table at v0, insert one row, and write a v1 parquet checkpoint.
+///
+/// Returns `(table_url, setup_engine, _temp_dir)` where `_temp_dir` must be kept
+/// alive for the duration of the test to prevent early cleanup. Callers can add
+/// further commits via `insert_data` or directly build the measuring snapshot.
+///
+/// Uses `TokioMultiThreadExecutor` because `snapshot.checkpoint()` issues nested
+/// `block_on` calls; `TokioBackgroundExecutor` deadlocks in that pattern.
+async fn setup_table_with_v1_checkpoint() -> DeltaResult<(
```
Comment (scottsand-db) [inline kernel/tests/metrics/main.rs:88]: Can you update this to use Drake's test builder stack? https://github.com/delta-io/delta-kernel-rs/pull/2320

File: kernel/tests/metrics/snapshot_load.rs [hunk 31]
```diff
@@ -0,0 +1,374 @@
...
+
+/// `snapshot.get_domain_metadata()` always performs a full log replay when no CRC is
+/// present at the target version. Calling it after a snapshot is already built generates
+/// a second round of JSON reads, demonstrating that on-demand metadata queries carry
+/// their own I/O cost.
+///
+/// The specific count (`json_read_calls = 1`) reflects this test's table: 2 commits, no
+/// checkpoint, no CRC. Tables with different log structures will produce different counts.
+#[tokio::test]
+async fn get_domain_metadata_incurs_additional_log_replay() -> DeltaResult<()> {
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs]: i think this test name should reflect `when_no_latest_crc`

File: kernel/tests/metrics/snapshot_load.rs [hunk 32]
```diff
@@ -0,0 +1,374 @@
...
+    );
+    assert!(reporter.json_files_read.get() > 0);
+
+    Ok(())
+}
+
+// ============================================================================
+// Scenario 9: snapshot.transaction() on a clustered table
+// ============================================================================
+// NOTE: This scenario is verified via clustering_e2e.rs tests because the
```
Comment (scottsand-db) [inline kernel/tests/metrics/snapshot_load.rs]: incomplete? todo? future?

Comment (chiinlquah) [reply kernel/tests/metrics/snapshot_load.rs]: removed, noise

File: kernel/tests/metrics/main.rs (hunk 30 above)
Comment (chiinlquah) [reply kernel/tests/metrics/main.rs:88]: Done, updated with merged helpers, left TODO for the unmerged helpers.
